File: app/src/main/java/com/krystianwsul/checkme/domainmodel/ObserverHolder.java
package com.krystianwsul.checkme.domainmodel;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.annimon.stream.Stream;
import com.krystianwsul.checkme.loaders.DomainLoader;

import junit.framework.Assert;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class ObserverHolder {
    @Nullable
    private static ObserverHolder sInstance;

    private final ArrayList<WeakReference<DomainLoader.Observer>> mObservers = new ArrayList<>();

    @NonNull
    public static synchronized ObserverHolder getObserverHolder() {
        if (sInstance == null)
            sInstance = new ObserverHolder();
        return sInstance;
    }

    public synchronized void addDomainObserver(@NonNull DomainLoader.Observer observer) {
        mObservers.add(new WeakReference<>(observer));
    }

    synchronized void clear() {
        mObservers.clear();
    }

    synchronized void notifyDomainObservers(@NonNull List<Integer> dataIds) {
        ArrayList<WeakReference<DomainLoader.Observer>> remove = new ArrayList<>();

        for (WeakReference<DomainLoader.Observer> reference : mObservers) {
            Assert.assertTrue(reference != null);

            DomainLoader.Observer observer = reference.get();
            if (observer == null)
                remove.add(reference);
            else
                observer.onDomainChanged(dataIds);
        }

        Stream.of(remove)
                .forEach(mObservers::remove);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/domainmodel/ObserverHolder.kt
package com.krystianwsul.checkme.domainmodel

import com.krystianwsul.checkme.viewmodels.DomainObserver
import java.lang.ref.WeakReference

object ObserverHolder {

    private val observers = mutableListOf<WeakReference<DomainObserver>>()

    @Synchronized
    fun addDomainObserver(observer: DomainObserver) = observers.add(WeakReference(observer))

    @Synchronized
    fun removeDomainObserver(observer: DomainObserver) {
        val remove = observers.filter { it.get().let { it == null || it == observer } }

        remove.forEach { observers.remove(it) }
    }

    @Synchronized
    fun notifyDomainObservers(dataIds: List<Int>) {
        val remove = observers.filter { it.get() == null }

        remove.forEach { observers.remove(it) }

        for (reference in observers)
            reference.get()!!.onDomainChanged(dataIds)
    }
}
